### PR TITLE
Final upstream vanilla changes

### PR DESCRIPTION
**Pull request by brianteeman**
* convert computeTextNodeWithImage helper function to js

* convert checkHeaders to JS

* convert computeAriaLabel to JS

### CODE REVIEW

File: dist/js/joomla-a11y-checker.umd.js [hunk 1]
```diff
@@ -1300,40 +1309,41 @@
                 <span class='sa11y-outline-list-item sa11y-yellow-text sa11y-bold'>${htext}</span>
             </li>`;
 
-                  if ($el.not(this.options.outlineIgnore).length !== 0) {
+              let ignoreArray = [];
+              if (sa11yOutlineIgnore) {
```
Comment (Fedik) [inline dist/js/joomla-a11y-checker.umd.js:1313]: such params in our code will be: `this.options.outlineIgnore`

Comment (brianteeman) [reply dist/js/joomla-a11y-checker.umd.js:1313]: yes I literally just checked that one

File: src/js/sa11y.js [hunk 2]
```diff
@@ -1294,40 +1303,41 @@ class Sa11y {
                 <span class='sa11y-outline-list-item sa11y-yellow-text sa11y-bold'>${htext}</span>
             </li>`;
 
-                if ($el.not(this.options.outlineIgnore).length !== 0) {
+            let ignoreArray = [];
+            if (outlineIgnore) {
```
Comment (Fedik) [inline src/js/sa11y.js]: I mean here :) 
such params in our code will be: `this.options.outlineIgnore`

File: src/js/sa11y.js [hunk 3]
```diff
@@ -1304,8 +1304,8 @@ class Sa11y {
             </li>`;
 
             let ignoreArray = [];
-            if (sa11yOutlineIgnore) {
-                ignoreArray = Array.from(document.querySelectorAll(sa11yOutlineIgnore));
+            if (outlineIgnore) {
```
Comment (brianteeman) [inline src/js/sa11y.js]: @fedik I am lost with js when something doesnt work.  Tried various things and I still get errors on this.

**Review by brianteeman — COMMENTED** [review]
```
Uncaught (in promise) ReferenceError: outlineIgnore is not defined
    at sa11y.js:1307
    at Array.forEach (<anonymous>)
    at Sa11y.checkHeaders (sa11y.js:1248)
    at checkAll (sa11y.js:731)
```

File: src/js/sa11y.js (hunk 3 above)
Comment (Fedik) [reply src/js/sa11y.js]: Change this section to:
```
if (this.options.outlineIgnore) {
  ignoreArray = Array.from(document.querySelectorAll(this.options.outlineIgnore));
}
```
Then should work

**Comment by Fedik** [comment]
All old parameters like `sa11yBlabla` and `sa11yOutlineIgnore` etc, now live in `this.options`.

`sa11yBlabla` become `this.options.blabla`
`sa11yOutlineIgnore` become `this.options.outlineIgnore`

**Comment by brianteeman** [comment]
I tried that.
```
            let ignoreArray = [];
            if (this.options.outlineIgnore) {
                ignoreArray = Array.from(document.querySelectorAll(outlineIgnore));
            }
```

But that just produced another error

```
sa11y.js:1316 Uncaught (in promise) TypeError: Cannot read properties of null (reading 'length')
    at sa11y.js:1316
    at Array.forEach (<anonymous>)
    at Sa11y.checkHeaders (sa11y.js:1248)
    at checkAll (sa11y.js:731)
```

File: src/js/sa11y.js [hunk 4]
```diff
@@ -1294,40 +1303,41 @@ class Sa11y {
...
+            if (!ignoreArray.includes(el)) {
                     //Append heading labels.
-                    $el.not(this.options.outlineIgnore).append(
-                        `<span class='sa11y-heading-label'>H${level}</span>`
-                    );
+                    el.insertAdjacentHTML("beforeend", `<span class='sa11y-heading-label'>H${level}</span>`);
 
                     //Heading errors
-                    if (error != null && $el.closest("a").length > 0) {
+                    if (error != null && el.closest("a").length != null) {
```
Comment (Fedik) [inline src/js/sa11y.js]: This cause a second error. In vanila js `.closest(blabla)` will return `null` if nothing found. Current code made in jquery style.

Change the IF condition to:
```
if (error != null && el.closest("a")) {
```

**Comment by Fedik** [comment]
I have added review-comment in the code, about second error

**Comment by brianteeman** [comment]
@Fedik I don;t see anything?

**Comment by Fedik** [comment]
In some reason a code reviews added at top only, scroll up ;)

**Comment by brianteeman** [comment]
That means you have not submitted the code review yet

**Comment by Fedik** [comment]
ah, now I get it, better use simple comments next time :smile:

**Comment by brianteeman** [comment]
Thanks @fedik - I had wrongly assumed that the first change had caused the second error. Instead it just revealed the error.

**Pull request merged**